### PR TITLE
mvcc: make schema upgrades gentle, old version of the PR

### LINKED ISSUE #2577: Schema of partition entries in cache and memtable is updated atomically, causing high latency and possible OOM on large partitions
After schema change, entries in memtables and cache are still using the old schema. They're upgraded lazily (but atomically) on access. Schema is currently tracked per-partition. This may badly impact latency if partitions are large.

### PR DESCRIPTION
**Pull request by michoecho**
After a schema change, memtable and cache have to be upgraded to the new schema. Currently, they are upgraded (on the first access after a schema change) atomically, i.e. all rows of the entry are upgraded with one non-preemptible call. This is a one of the last vestiges of the times when partition were treated atomically, and it is a well known source of numerous large stalls.

This series makes schema upgrades gentle (preemptible). This is done by co-opting the existing MVCC machinery.
Before the series, all partition_versions in the partition_entry chain have the same schema, and an entry upgrade replaces the entire chain with a single squashed and upgraded version.
After the series, each partition_version has its own schema. A partition entry upgrade happens simply by adding an empty version with the new schema to the head of the chain. Row entries are upgraded to the current schema on-the-fly by the cursor during reads, and upgraded permanently by the MVCC version merge ongoing in the background.

The series:
1. Does some code cleanup in the mutation_partition area.
2. Adds a schema field to partition_version and removes it from its containers (partition_snapshot, cache_entry, memtable_entry).
3. Adds upgrading variants of constructors and apply() for `row` and its wrappers. 
4. Prepares partition_snapshot_row_cursor, mutation_partition_v2::apply_monotonically and partition_snapshot::merge_partition_versions for dealing with heterogeneous version chains.
5. Modifies partition_entry::upgrade to perform upgrades by extending the version chain with a new schema instead of squashing it to a single upgraded version. 

I'm particularly worried about the correctness of the modifications to the version merging algorithm. It is already probably the trickiest and most complex piece of Scylla code, and these modifications make reasoning even harder.
Before the series, we always merge the newer version into the older version. After the series, we also merge in the other direction, because after the merge we want to end up with the newer schema, not the older schema. This changes some assumptions of apply_monotonically, especially these concerning eviction order and the "older versions are evicted first" rule.
Between preemption safety, exception safety, rows and dummies and last dummies, evictable and non-evictable entries, continuity, row tombstones, row compaction, and now also forward and backward merging, the complexity of this function is brain-melting. I have no idea how to cover it with trustworthy tests. The puny test I've added is barely more than a sanity check.

A possible performance problem with this method is that the same row entry might have to be upgraded many times by the cursor until its finally upgraded permanently by a version merge, whereas before the series they only need to be upgraded once. Note that problem is only relevant for big partition entries (the version merge starts eagerly and is only backgrounded on first preempt, so for small entries it will usually complete synchronously). I didn't do any performance tests, but schema upgrades currently (both before and after this series) seem to be implemented inefficiently: for *every cell* in the upgraded entry, a schema upgrade looks up its column name in a hashmap. If this is a performance problem, we can follow up with an optimization which avoids this by preparing an integer-based column mapping for the schema pair instead.

Fixes #2577

### CODE REVIEW

**Comment by michoecho** [comment]
Ready for review. Requesting review from @tgrabiec

**Comment by michoecho** [comment]
v1.1
Debug mode CI found a real bug. partition_version now possesses a `schema_ptr`, so a destruction of `partition_version` can lead to a destruction of `schema`. However, schema can contain a LSA type (e.g. `managed_bytes`, via `view_info`->`select_statement`->`expression::impl`->`binary_operator`->`constant`->`raw_value`), so it wrongly destroyed in `partition_version`'s allocator, instead of `schema`'s allocator.
I fixed this in the PR by modifying the destructor of `partition_version` to nullify the schema_ptr in `standard_allocator()`.

I think that the existing code is also exposed to this bug, though. (Except the problem is with the destruction of `cache_entry`/`memtable_entry` instead of `partition_version`).
@tgrabiec What do you think? What's the lifecycle of the `schema` object? I guess that if `cache_entry` stores a `schema_ptr`, then it's possible for a destruction of `cache_entry` to destroy the `schema`?

**Comment by michoecho** [comment]
v1.0.2
dtests found another bug. I missed one place in the cursor code (the argument to prepare_hash(), which is only called when `digest_requested` — that's why most tests miss it) which should have been changed from using the cursor's schema to using the version's schema. Fixed.

**Comment by michoecho** [comment]
CI currently failing due to some irrelevant stuff: 
```
20:02:18  + docker system prune -a -f --filter until=72h
20:02:18  Error response from daemon: a prune operation is already running
```

**Comment by avikivity** [comment]
@benipeled please help with CI going wild

**Comment by michoecho** [comment]
> @benipeled please help with CI going wild

He already helped, there was a disk space issue, but it should be fixed now.
I now started a new run at https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/25/console

**Comment by tgrabiec** [comment]
On Tue, Mar 14, 2023 at 2:13 PM Michał Chojnowski ***@***.***>
wrote:

> v1.1
> Debug mode CI found a real bug. partition_version now possesses a
> schema_ptr, so a destruction of partition_version can lead to a
> destruction of schema. However, schema can contain a LSA type (e.g.
> managed_bytes, via view_info->select_statement->expression::impl->
> binary_operator->constant->raw_value), so it wrongly destroyed in
> partition_version's allocator, instead of schema's allocator.
> I fixed this in the PR by modifying the destructor of partition_version
> to nullify the schema_ptr in standard_allocator().
>
> I think that the existing code is also exposed to this bug, though.
> (Except the problem is with the destruction of cache_entry/memtable_entry
> instead of partition_version).
> @tgrabiec <https://github.com/tgrabiec> What do you think? What's the
> lifecycle of the schema object? I guess that if cache_entry stores a
> schema_ptr, then it's possible for a destruction of cache_entry to
> destroy the schema?
>
Yes, cache_entry can destroy the schema.

schema is not supposed to be a managed object, so we don't care about
allocator context. But it contains a managed object (raw_value), and that's
a problem. raw_value should be switched to be non-managed. If we have no
alternative for managed_bytes, we can create a wrapper which destroys it in
std allocator context.

File: test/lib/flat_mutation_reader_assertions.hh [hunk 1]
```diff
@@ -504,7 +504,7 @@ public:
     mutation_assertion next_mutation() {
         auto mo = read_mutation_from_flat_mutation_reader(_reader).get0();
         BOOST_REQUIRE(bool(mo));
-        return mutation_assertion(std::move(*mo));
+        return assert_that(std::move(*mo));
```
[the remark this answers is not in the document]
Comment (tgrabiec) [reply test/lib/flat_mutation_reader_assertions.hh:507]: https://github.com/scylladb/scylladb/issues/12073

**Comment by michoecho** [comment]
> On Tue, Mar 14, 2023 at 2:13 PM Michał Chojnowski ***@***.***> wrote: v1.1 Debug mode CI found a real bug. partition_version now possesses a schema_ptr, so a destruction of partition_version can lead to a destruction of schema. However, schema can contain a LSA type (e.g. managed_bytes, via view_info->select_statement->expression::impl-> binary_operator->constant->raw_value), so it wrongly destroyed in partition_version's allocator, instead of schema's allocator. I fixed this in the PR by modifying the destructor of partition_version to nullify the schema_ptr in standard_allocator(). I think that the existing code is also exposed to this bug, though. (Except the problem is with the destruction of cache_entry/memtable_entry instead of partition_version). @tgrabiec <https://github.com/tgrabiec> What do you think? What's the lifecycle of the schema object? I guess that if cache_entry stores a schema_ptr, then it's possible for a destruction of cache_entry to destroy the schema?
> Yes, cache_entry can destroy the schema. schema is not supposed to be a managed object, so we don't care about allocator context. But it contains a managed object (raw_value), and that's a problem. raw_value should be switched to be non-managed. If we have no alternative for managed_bytes, we can create a wrapper which destroys it in std allocator context.
> […](#)
> Message ID: ***@***.***>

@avikivity Using managed_bytes as the common type for the buffer fragmentation project was a mistake.

File: mutation/mutation_partition_v2.cc [hunk 2]
```diff
@@ -114,17 +113,27 @@ std::ostream& operator<<(std::ostream& out, const apply_resume& res) {
...
 }
 
-stop_iteration mutation_partition_v2::apply_monotonically(const schema& s, mutation_partition_v2&& p, cache_tracker* tracker,
-        mutation_application_stats& app_stats, is_preemptible preemptible, apply_resume& res, is_evictable evictable) {
-    return apply_monotonically(s, std::move(p), tracker, app_stats,
-        preemptible ? default_preemption_check() : never_preempt(), res, evictable);
+void mutation_partition_v2::apply_v1(const schema& s, mutation_partition&& p) {
+    apply_v2(s, mutation_partition_v2(s, std::move(p)));
+}
+void mutation_partition_v2::apply_v2(const schema& s, mutation_partition_v2&& p, cache_tracker* tracker, is_evictable evictable) {
```
[the remark this answers is not in the document]
Comment (michoecho) [reply mutation/mutation_partition_v2.cc:119]: In this case, personal preference. Overloads break grepping for the name, so I don't like giving the same name to two functions without a reason.

File: test/boost/mutation_test.cc [hunk 3]
```diff
@@ -2101,7 +2101,7 @@ SEASTAR_TEST_CASE(test_v2_merging_in_non_evictable_snapshot) {
             auto to_apply = mutation_partition_v2(s, m2_v2);
 
             apply_resume res;
-            while (result_v2.apply_monotonically(s, std::move(to_apply), no_cache_tracker, app_stats,
+            while (result_v2.apply_monotonically(s, s, std::move(to_apply), no_cache_tracker, app_stats,
```
[the remark this answers is not in the document]
Comment (michoecho) [reply test/boost/mutation_test.cc:2104]: You are trading 10 chars of clutter in the test for >4 lines of clutter in `mutation_partition_v2.{hh,cc}`. IMO the helper wasn't carrying its own weight.

File: mutation/mutation_partition.hh [hunk 4]
```diff
@@ -851,11 +849,8 @@ public:
         _deleted_at.maybe_shadow(_marker);
     }
 
-    // Weak exception guarantees. After exception, both src and this will commute to the same value as
-    // they would should the exception not happen.
+    // Weak exception guarantees.
     void apply(const schema& s, const deletable_row& src);
-    void apply(const schema& s, deletable_row&& src);
-    void apply_monotonically(const schema& s, const deletable_row& src);
     void apply_monotonically(const schema& s, deletable_row&& src);
```
[the remark this answers is not in the document]
Comment (michoecho) [reply mutation/mutation_partition.hh:870]: I see that I made it seem that the "weak exception guarantees" applies to `apply_monotonically`, when it applies only to `apply`. I will correct that.

That said, the guarantee you mentioned could apply to `apply` too, I think. It already does in practice, and we don't have a good reason to keep it relaxed. It's quite hard to write a non-monotonic variant of this.

File: mutation/mutation_fragment.hh [hunk 5]
```diff
@@ -67,10 +67,10 @@ public:
     }
 
     void apply(const schema& s, clustering_row&& cr) {
-        _row.apply(s, std::move(cr._row));
+        _row.apply_monotonically(s, std::move(cr._row));
```
[the remark this answers is not in the document]
Comment (michoecho) [reply mutation/mutation_fragment.hh:70]: I understand the value of giving callers as weak guarantees as possible, but you have to weight that against having more variants. If we invent an artificial variant just to weaken guarantees, that's just confusing. I find it hard to imagine any performance benefit to a non-monotonic variant. Do you have something specific in mind?

Besides, I'm modifying the implementation here, not the interface. There's nothing wrong with using a stronger variant in the implementation. It doesn't state anything. The monotonical variant is more fundamental anyway, so it's not like I'm adding dependencies either.

File: mutation/mutation_partition.hh [hunk 6]
```diff
@@ -178,6 +178,7 @@ public:
 
     // Weak exception guarantees
     void apply(const schema&, column_kind, const row& src);
+    void apply(const schema& from, const schema& to, column_kind kind, const row& other);
```
[the remark this answers is not in the document]
Comment (michoecho) [reply mutation/mutation_partition.hh:182]: Good catch. I meant to rename these but I forgot.

File: test/lib/flat_mutation_reader_assertions.hh (hunk 1 above)
[the remark this answers is not in the document]
Comment (michoecho) [reply test/lib/flat_mutation_reader_assertions.hh:507]: Right, maybe it really is a fool's errand. I will drop this patch.

File: test/boost/row_cache_test.cc [hunk 7]
```diff
@@ -3448,6 +3448,16 @@ SEASTAR_TEST_CASE(test_concurrent_reads_and_eviction) {
...
+            bool upgrade_schema = tests::random::get_bool();
+            if (upgrade_schema) {
+                schema_ptr new_schema = schema_builder(s)
+                    .with_column(to_bytes("_phantom"), byte_type)
+                    .remove_column("_phantom")
+                    .build();
+                m2.upgrade(new_schema);
+                cache.set_schema(new_schema);
+            }
+
```
Comment (avikivity) [inline test/boost/row_cache_test.cc:3461]: Regarding tests, I think we need a stresser that builds a model of the data outside the row cache and compares it with the real row_cache under a mix of writes, reads, schema changes, and evictions. The snapshot_source that backs the cache can read the model, no need for real sstables.

**Comment by avikivity** [comment]
> After a schema change, memtable and cache have to be upgraded to the new schema. Currently, they are upgraded (on the first access after a schema change) atomically, i.e. all rows of the entry are upgraded with one non-preemptible call. This is a one of the last vestiges of the times when partition were treated atomically, and it is a well known source of numerous large stalls.
> 
> This series makes schema upgrades gentle (preemptible). This is done by co-opting the existing MVCC machinery. Before the series, all partition_versions in the partition_entry chain have the same schema, and an entry upgrade replaces the entire chain with a single squashed and upgraded version. After the series, each partition_version has its own schema. A partition entry upgrade happens simply by adding an empty version with the new schema to the head of the chain. Row entries are upgraded to the current schema on-the-fly by the cursor during reads, and upgraded permanently by the MVCC version merge ongoing in the background.
> 
> The series:
> 
> 1. Does some code cleanup in the mutation_partition area.
> 2. Adds a schema field to partition_version and removes it from its containers (partition_snapshot, cache_entry, memtable_entry).

This is neutral in terms of non-transient memory footprint, yes?


> 3. Adds upgrading variants of constructors and apply() for `row` and its wrappers.
> 4. Prepares partition_snapshot_row_cursor, mutation_partition_v2::apply_monotonically and partition_snapshot::merge_partition_versions for dealing with heterogeneous version chains.
> 5. Modifies partition_entry::upgrade to perform upgrades by extending the version chain with a new schema instead of squashing it to a single upgraded version.
> 
> I'm particularly worried about the correctness of the modifications to the version merging algorithm. It is already probably the trickiest and most complex piece of Scylla code, and these modifications make reasoning even harder. Before the series, we always merge the newer version into the older version. After the series, we also merge in the other direction, because after the merge we want to end up with the newer schema, not the older schema.

Could we not make things more orthogonal? For example, instead of an upgrading merge, create a clone of the version with the new schema, replace the old version with the upgraded schema, and then proceed with the version merge?


A different way to make things orthogonal is to create a row_cache object for every schema version and chain those row_cache objects, but that comes with its own problems. I admit the motivation for that was to avoid duplicating the schema everywhere, not a desire to simplify things.

> This changes some assumptions of apply_monotonically, especially these concerning eviction order and the "older versions are evicted first" rule. Between preemption safety, exception safety, rows and dummies and last dummies, evictable and non-evictable entries, continuity, row tombstones, row compaction, and now also forward and backward merging, the complexity of this function is brain-melting. I have no idea how to cover it with trustworthy tests. The puny test I've added is barely more than a sanity check.
> 
> A possible performance problem with this method is that the same row entry might have to be upgraded many times by the cursor until its finally upgraded permanently by a version merge, whereas before the series they only need to be upgraded once. 

I think it's reasonable, as long as we schedule the version merge to happen "soon".

> Note that problem is only relevant for big partition entries (the version merge starts eagerly and is only backgrounded on first preempt, so for small entries it will usually complete synchronously).

That's fine. For small partitions we get good behavior, and for large partitions anything is better than the stall we have today.

> I didn't do any performance tests, but schema upgrades currently (both before and after this series) seem to be implemented inefficiently: for _every cell_ in the upgraded entry, a schema upgrade looks up its column name in a hashmap. If this is a performance problem, we can follow up with an optimization which avoids this by preparing an integer-based column mapping for the schema pair instead.
> 
> Fixes #2577

Please annotate the cover letter with perf-simple-query results.

**Comment by tgrabiec** [comment]
On Wed, Mar 15, 2023 at 2:52 PM Michał Chojnowski ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In mutation/mutation_partition_v2.cc
> <https://github.com/scylladb/scylladb/pull/13104#discussion_r1137111359>:
>
> > @@ -114,17 +113,27 @@ std::ostream& operator<<(std::ostream& out, const apply_resume& res) {
>      return out << "{" << int(res._stage) << ", " << res._pos << "}";
>  }
>
> -stop_iteration mutation_partition_v2::apply_monotonically(const schema& s, mutation_partition_v2&& p, cache_tracker* tracker,
> -        mutation_application_stats& app_stats, is_preemptible preemptible, apply_resume& res, is_evictable evictable) {
> -    return apply_monotonically(s, std::move(p), tracker, app_stats,
> -        preemptible ? default_preemption_check() : never_preempt(), res, evictable);
> +void mutation_partition_v2::apply_v1(const schema& s, mutation_partition&& p) {
> +    apply_v2(s, mutation_partition_v2(s, std::move(p)));
> +}
> +void mutation_partition_v2::apply_v2(const schema& s, mutation_partition_v2&& p, cache_tracker* tracker, is_evictable evictable) {
>
> In this case, personal preference. Overloads break grepping for the name,
> so I don't like giving the same name to two functions without a reason.
>

This code will be read more often than grepped. In this case it looks odd
because mutation_partition_v2 is the default (same object). v1 can be
justified.

**Comment by tgrabiec** [comment]
On Wed, Mar 15, 2023 at 2:54 PM Michał Chojnowski ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In test/boost/mutation_test.cc
> <https://github.com/scylladb/scylladb/pull/13104#discussion_r1137114509>:
>
> > @@ -2101,7 +2101,7 @@ SEASTAR_TEST_CASE(test_v2_merging_in_non_evictable_snapshot) {
>              auto to_apply = mutation_partition_v2(s, m2_v2);
>
>              apply_resume res;
> -            while (result_v2.apply_monotonically(s, std::move(to_apply), no_cache_tracker, app_stats,
> +            while (result_v2.apply_monotonically(s, s, std::move(to_apply), no_cache_tracker, app_stats,
>
> You are trading 10 chars of clutter in the test for >4 lines of clutter in
> mutation_partition_v2.{hh,cc}. IMO the helper wasn't carrying its own
> weight.
>

The clutter in tests weighs more because it's what you look at when trying
to understand the failed test. Pushing clutter down is good.

**Comment by michoecho** [comment]
>> The series:
>> Does some code cleanup in the mutation_partition area.
>> Adds a schema field to partition_version and removes it from its containers (partition_snapshot, cache_entry, memtable_entry).
>
> This is neutral in terms of non-transient memory footprint, yes?

Yes.

> Could we not make things more orthogonal? For example, instead of an upgrading merge, create a clone of the version with the new schema, replace the old version with the upgraded schema, and then proceed with the version merge?

We could. It's on the table. In this scenario, we could have a simpler "upgrading merge" function, upgrade_monotonically(), which would be used only for heterogeneous neighbors, and only if the two versions cover disjoint ranges — it would be simpler, because the function would only need to move entries and insert sentinels on preempt, but wouldn't have to care about intersections, which are the hard part. (And if there were two heterogeneous versions which don't cover disjoint ranges, then an empty version with the newer schema would be inserted between them to enforce that).

I can try writing that variant too. Lowering the complexity of `apply_monotonically` can be worth it.

> I think it's reasonable, as long as we schedule the version merge to happen "soon".

Version merge is scheduled as soon as the snapshot which owns the version is destroyed, but version merges happen in order (except for the "first task quota" of the merge, which happens synchronously with the destruction, until it's preempted). The merges happen in a special scheduling group, `mem_compaction`. (Not making a point, just reminding how it works).

> Please annotate the cover letter with perf-simple-query results.

Will do.

**Comment by michoecho** [comment]
> This code will be read more often than grepped.

In my case grepping is an essential part of reading. If I'm looking at some code, usually I also want to know where it's used, so I grep for its calls. If the name repeats too much, you need to fire up an IDE to find the call site, and I don't like that.

But since you insist, I will rename them back to just `apply()`.

> The clutter in tests weighs more because it's what you look at when trying
to understand the failed test. Pushing clutter down is good.

Are you saying that pushing clutter from test code to core code is good? I disagree with the sentiment, IMHO readability of core code takes priority over readability of tests. But I can add the helper back.

**Comment by avikivity** [comment]
There is a third option yet, flattening the versions. In this option we allow rows with duplicate clustering keys (and duplicate static rows), the duplicates are disambiguated by the maximum timestamp within the row's mutation. A version is an instruction not to merge rows across a particular timestamp. Reading a version means reading the partition and ignoring higher-timestamp rows. To remove a version, merge adjacent rows that cross the version's timestamp.

It only works for LWT and raft (since monotonic timestamps are required), but MVCC is useless anyway outside LWT/Raft.

**Comment by michoecho** [comment]
CI failed on the ol' reliable ttl_test.TestDistributedTTL.test_ttl_is_respected_on_repair

Restarting. This time it will pass, I feel it.

**Comment by avikivity** [comment]
> CI failed on the ol' faithful ttl_test.TestDistributedTTL.test_ttl_is_respected_on_repair
> 
> Restarting. This time it will pass, I can feel it.

Was an issue filed/updated?

**Comment by michoecho** [comment]
> Was an issue filed/updated?

It's https://github.com/scylladb/scylladb/issues/11684. It's failing some CI every day.

**Comment by avikivity** [comment]
Asked to prioritize.

File: mutation/mutation_partition.hh [hunk 8]
```diff
@@ -97,6 +97,7 @@ public:
     row();
     ~row();
     row(const schema&, column_kind, const row&);
+    static row construct_row(const schema& our_schema, const schema& their_schema, column_kind, const row&);
```
[the remark this answers is not in the document]
Comment (michoecho) [reply mutation/mutation_partition.hh:100]: `construct_row` is greppable, `construct` is not.
In the absence of other arguments, I prefer the more greppable name.

File: mutation/mutation_partition_v2.cc [hunk 9]
```diff
@@ -440,18 +447,26 @@ stop_iteration mutation_partition_v2::apply_monotonically(const schema& s, const
                 }
```
[the remark this answers is not in the document]
Comment (michoecho) [reply mutation/mutation_partition_v2.cc:447]: Ah, so this is what I had in mind when I changed s2 to `p_s`.

File: mutation/mutation_partition_v2.cc [hunk 10]
```diff
@@ -440,18 +447,26 @@ stop_iteration mutation_partition_v2::apply_monotonically(const schema& s, const
...
-                // the last dummy entry which holds the partition entry linked in LRU.
-                i->swap(src_e);
+                if (same_schema) [[likely]] {
+                    // Newer evictable versions store complete rows
+                    i->row() = std::move(src_e.row());
+                    // Need to preserve the LRU link of the later version in case it's
+                    // the last dummy entry which holds the partition entry linked in LRU.
+                    i->swap(src_e);
+                } else {
+                    i->apply_monotonically(s, p_s, std::move(src_e));
```
[the remark this answers is not in the document]
Comment (michoecho) [reply mutation/mutation_partition_v2.cc:457]: True.

File: mutation/mutation_fragment.hh (hunk 5 above)
[the remark this answers is not in the document]
Comment (tgrabiec) [reply mutation/mutation_fragment.hh:70]: > Do you have something specific in mind?

For example, non-monotonic variant can remove items from src without rebalancing the src tree, while monotonic variant has to keep src consistent and rebalance it. For schema-upgrading apply, if we don't yet have the on-the-fly algorithm, you can't upgrade src in-place and then apply, you have to copy it first. 

At some point we can decide that there's no reason to have two variants, but as long as we do, we shouldn't change users of apply() to apply_monotonically().

File: mutation/mutation_partition.hh (hunk 8 above)
[the remark this answers is not in the document]
Comment (michoecho) [reply mutation/mutation_partition.hh:100]: But from within the implementation of `class row` you can call it without the `row::` scope — just `construct()`. So if you really want to grep all uses of a function in C++, you have to grep just its name, not the qualified name. (And there are also function pointers, and typedefs, and more). A static method with a generic method is more greppable than a non-static with a generic name, but not truly greppable.

[the remark this answers is not in the document]
Comment (tgrabiec) [reply mutation/mutation_partition.hh:100]: True. But should we go there? And rename row::size() to row::row_size() too, and apply() to mpv2_apply()? It seems like a slippery slope to me, which hurts readability. This problem is solved by IDE's with "Find Usages". @avikivity What's your take?

**Comment by tgrabiec** [comment]
> @avikivity Using managed_bytes as the common type for the buffer fragmentation project was a mistake.

Please open an issue so that we don't forget about it.

File: mutation/mutation_partition.hh (hunk 8 above)
[the remark this answers is not in the document]
Comment (michoecho) [reply mutation/mutation_partition.hh:100]: > This problem is solved by IDE's with "Find Usages".

It is solved by IDEs until it isn't.

For example:
https://github.com/scylladb/scylladb/blob/5705df77a15531580da22ba1e78614793b51352c/serializer.hh#L254-L257

clangd seems to have no idea where are the implementations of write(), even though this is statically resolvable information.
If it had a unique name, I could at least grep for it — but I can't.

The problem goes the other way too: the IDE won't show you a usage of a method if it's used from a type substituted into a template.

It's not like I like it — I don't think greppability is "the right thing". If "find references" actually worked, I wouldn't advocate it. But C++ tooling was retrofitted onto the language and is less reliable than plain text.

**Comment by michoecho** [comment]
> > @avikivity Using managed_bytes as the common type for the buffer fragmentation project was a mistake.
> 
> Please open an issue so that we don't forget about it.

Opened https://github.com/scylladb/scylladb/issues/13216

File: mutation/partition_version.cc [hunk 11]
```diff
@@ -201,23 +201,36 @@ stop_iteration partition_snapshot::merge_partition_versions(mutation_application
...
+                bool unique_owner = prev_ref.is_unique_owner();
+                merged_into->back_reference().release();
+                prev_ref = partition_version_ref(*merged_into, unique_owner);
+                current_allocator().destroy(merged_from);
                 return stop_iteration::yes;
             }
-            current_allocator().destroy(prev);
+            if (merged_into->is_referenced()) {
+                merged_into->back_reference().release();
+            }
```
[the remark this answers is not in the document]
Comment (michoecho) [reply mutation/partition_version.cc:231]: The line below creates a reference to merged_into, and it's forbidden to create a reference to an already-referenced version. (This is enforced by this assert:)
https://github.com/scylladb/scylladb/blob/aad2afd417302c0405b07781f4bf1bfeca7b84f7/mutation/partition_version.hh#L158

So we have to release the current reference to merged_into explicitly.
But I concur that this looks wrong. I'll add a comment which explains it.

**Comment by avikivity** [comment]
Let's move this forward. What's remaining?

**Comment by michoecho** [comment]
> Let's move this forward. What's remaining?

Status reminder:

The PR aimed to implement gentle upgrades by making the schema a per-version property (instead of a per-entry property), and adding support for heterogeneous entries to reading and merging algorithms. This way an upgrade boils down to adding an empty version.
But this requires merging versions forward (moving rows from an older version into a newer version), which is the opposite of the regular merging direction. We ultimately decided that requiring the merge algorithm to work in both directions makes it too hard to comprehend (and believe in its correctness).
Tomek gave an alternative idea in https://github.com/scylladb/scylladb/pull/13104#issuecomment-1471912729, but it's not that simple either. It avoids forward merges, but it adds much complexity to how the version chain is managed.

So, in short, progress was stalled mostly because no design seems good, and we (I) are afraid of pushing yet another bunch of complexity into an already very complicated section of code.

I was just planning to rewrite this into a variant of Tomek's idea (where during upgrade we add a newer version after the older version, and the two are treated from the outside as a unit, and they aren't merged from/to until their merge/upgrade finishes), but the tracing project (and other issues) diverted my time.

A secondary unsolved problem (https://github.com/scylladb/scylladb/pull/13104#discussion_r1138834196) is that an assumption made by the PR — that a merge can be triggered on upgrade to make upgrades of small partition entries effectively synchronous — was wrong, because of allocating_section nesting rules (there can be no nesting). Since both the cache update and the merge have their own allocating section, the merge refuses to happen during the update, so it becomes asynchronous, rather than synchronous. We have to avoid this somehow, maybe by bypassing the merge section if we are inside the update section.

**Comment by avikivity** [comment]
@tgrabiec and @michoecho, please agree on a direction and move forward, it's a pity to let this rot.

**Comment by mykaul** [comment]
> @tgrabiec and @michoecho, please agree on a direction and move forward, it's a pity to let this rot.

@tgrabiec , @michoecho - please move forward, I'd be happy to see it in before 5.3 branch.

**Comment by michoecho** [comment]
> > @tgrabiec and @michoecho, please agree on a direction and move forward, it's a pity to let this rot.
> 
> @tgrabiec , @michoecho - please move forward, I'd be happy to see it in before 5.3 branch.

Note: a new version of this PR is at https://github.com/scylladb/scylladb/pull/13761. It's awaiting @tgrabiec's review.

**Comment by avikivity** [comment]
Closing in favor of #13761 to avoid confusion.

**Pull request closed without merging**